### ISSUE TITLE
SMS Streams recovery

### ISSUE THREAD

**Issue by redss** [dotnet/orleans#1995]
Hello,

We have two services:
1. Orleans silo and
2. client service - a Windows Service hosting SignalR.

Client service is meant to receive notifications from silo and push them further to SignalR. We decided to work with SMS streams: some grains are publishing messages onto the stream, and client service is consuming them.

Everything works fine, until we restart the silo. After that, we can see that GrainClient reconnects, but client service no longer receives any data from the stream. Is such behavior expected of SMS streams? If it is, should we use other stream providers in order to have more resilient connection?

**Comment by gabikliot** [comment]
Are you using persistent storage provider for pub sub?

**Comment by redss** [comment]
Yes, we're using AzureStoreageProvider for pub sub.

I've also noticed, that we get following warnings on silo servcie:

```
2016-08-01 13:36:19 [Warning] SimpleMessageStreamProducerExtension: Consumer ObserverId:*cli/4f944baa/b6e80465 on stream AreaOccupancyChanged-00000000-0000-0000-0000-000000000000-StreamProvider is no longer active - permanently removing Consumer..
```

After getting such warning, client service no longer receives any messages. I know, that grains have to explicitly resume their streams, when they are reactivated. But do we have to explicitly resume streams in client code?

**Comment by jason-bragg** [comment]
@redss.  I apologize for the delayed response.  A client should not be required to re-subscribe due to silo crash.  What version of Orleans are you using?

**Comment by redss** [comment]
We're using Orleans 1.2.3.

**Comment by jason-bragg** [comment]
Trying to reproduce this in test and have some questions.

"Everything works fine, until we restart the silo."
Silo?  You have a single silo?

"After that, we can see that GrainClient reconnects"
Reconnects, but does not resubscribe, correct?

Can you try running two silos in the cluster.

If a client loses connection to the entire cluster (which will happen whenever a cluster has 1 host and it dies), the streaming infrastructure can't recover, but it should recover as long as the client is still connected to the cluster, even if  a single silo dies.

For streaming, if a client looses connection to the entire cluster, it will need to resubscribe once it reconnects to the cluster.

**Comment by redss** [comment]
Hello,

"Silo? You have a single silo?"
Yes, we were testing streams on a single silo.

"Reconnects, but does not resubscribe, correct?"
Yes.

Ok, so streams cannot recover after losing connection to a cluster. I think this answers my initial question and explains our case very well. However, I still think we should be ready for such incident. Do you have any patterns for that? I couldn't find any means in OrleansClient to subscribe to events like being reconnected to a cluster (so then I could resubscribe my streams).

**Comment by sergeybykov** [comment]
>Ok, so streams cannot recover after losing connection to a cluster. I think this answers my initial question and explains our case very well.

That's not what @jason-bragg said:
>If a client loses connection to the entire cluster (which will happen whenever a cluster has 1 host and it dies), the streaming infrastructure can't recover, but it should recover as long as the client is still connected to the cluster, even if a single silo dies.

A client will not auto-recover after it loses connection to *every* silo of the cluster. So long as it is connected, to at least one silo, it should be fine.

Streams in Orleans are virtual, they don't ever need to recover. Consumer may need to re-subscribe in certain situations like the above case of a client losing connection to the cluster.

**Comment by jason-bragg** [comment]
@redss, As Sergey stated, there is currently no notification to the application layer that a client has lost connection to the cluster.  We've discussed this behavior but there has never been sufficient need.  Most of our customers run sufficiently high numbers of silos that losing connectivity to the cluster is virtually non-existent.  If you've need of this feature we'd, of course, be happy to review a PR.

A more application specific solution could be to introduce some sort of periodic ping from the stream producer.  Some systems refer to this as a 'punctuation'.  If a grain goes longer than a set period of time without generating an event, it generates a punctuation event telling the client that it's still around and that the stream is still active.  With this punctuation event, any client would be sure to receive an event, real, or punctuation, periodically.  If the client does not see an event for more than some multiple of the punctuation period, it can be confident it's lost connectivity to the cluster, and should restart.

**Comment by gabikliot** [comment]
@jason-bragg , I would like to make sure you are not actually suggesting customers invest in somewhat complicated work arounds for what we consider a bug in the streaming infrastructure.
 
> If a client loses connection to the entire cluster (which will happen whenever a cluster has 1 host and it dies), the streaming infrastructure can't recover

Yes, in certain scenarios it cannot recover and it is a known BUG which we should fix. We know we can fix it without any "notification to the application layer that a client has lost connection to the cluster".

**Comment by jason-bragg** [comment]
@gabikliot I know very little about @redss's needs, or resources.  I know only that @redss seeks remedy and this bug has not been fixed due to low priority, so I'm offering some options to unblock @redss.  Yes, one of those options is a workaround.

I challenge the framing of @redss as a customer, rather than a community member or consumer.  As an open source initiative, what is important enough to the community to get addressed, usually gets addressed by the community.  The vender/customer relationship, imo, is not applicable to open source projects.

As for fixes without notifying client of the connection loss, the notification is, IMO, a minimal fix and does not preclude the development of more involved solutions if/when needed.  I am not opposed to more involved solutions, but suspect a simpler solution would be more likely to be implemented.
